internal/controller/controllers/bmh_agent_controller.go

```diff
@@ -0,0 +1,499 @@
+			SizeBytes:    bmh_v1alpha1.Capacity(d.SizeBytes),
+			SerialNumber: d.Serial,
+			WWN:          d.Wwn,
+		}
+
+		hardwareDetails.Storage = append(hardwareDetails.Storage, disk)
+	}
+
+	// Add the memory information in MebiByte
+	if agent.Status.Inventory.Memory.PhysicalBytes > 0 {
+		hardwareDetails.RAMMebibytes = int(inventory.Memory.PhysicalBytes * 1024 * 1024)
+	}
+
+	// Add CPU information
+	hardwareDetails.CPU = bmh_v1alpha1.CPU{
+		Flags:          inventory.Cpu.Flags,
+		Count:          int(inventory.Cpu.Count),
+		Model:          inventory.Cpu.ModelName,
+		Arch:           inventory.Cpu.Architecture,
+		ClockMegahertz: bmh_v1alpha1.ClockSpeed(inventory.Cpu.ClockMegahertz),
+	}
+
+	// BMH has a `virtual` field that the Agent does not have
+	hardwareDetails.SystemVendor = bmh_v1alpha1.HardwareSystemVendor{
+		Manufacturer: inventory.SystemVendor.Manufacturer,
+		ProductName:  inventory.SystemVendor.ProductName,
+		SerialNumber: inventory.SystemVendor.SerialNumber,
+	}
+
+	bytes, err := json.Marshal(hardwareDetails)
+	if err != nil {
+		return reconcileError{err}
+	}
+
+	if bmh.ObjectMeta.Annotations == nil {
+		bmh.ObjectMeta.Annotations = make(map[string]string)
+	}
+
+	bmh.ObjectMeta.Annotations[BMH_HARDWARE_DETAILS_ANNOTATION] = string(bytes)
+	return reconcileComplete{true}
+
+}
+
+// Reconcile the `BareMetalHost` resource
+//
+// This reconcile step sets the Image.URL value in the `BareMetalHost`
+// spec by copying it from the `InstallEnv` referenced in the resource's
+// labels. If the previous action succeeds, this step will also set the
+// BMH_INSPECT_ANNOTATION to disabled on the BareMetalHost.
+//
+// The above changes will be done only if the ISODownloadURL value has already
+// been set in the `InstallEnv` resource and the Image.URL value has not been
+// set in the `BareMetalHost`
+func (r *BMACReconciler) reconcileBMH(bmh *bmh_v1alpha1.BareMetalHost) reconcileResult {
+	ctx := context.Background()
+
+	// No need to reconcile if the image URL has been set in
+	// the BMH already.
+	if bmh.Spec.Image != nil && bmh.Spec.Image.URL != "" {
+		return reconcileComplete{}
+	}
+
+	r.Log.Debugf("Started BMH reconcile for %s/%s", bmh.Namespace, bmh.Name)
+
+	for ann, value := range bmh.Labels {
+
+		// Find the `BMH_INSTALL_ENV_LABEL`, get the isntallEnv configured in it
+		// and copy the ISO Url from the InstallEnv to the BMH resource.
+		if strings.HasPrefix(ann, BMH_INSTALL_ENV_LABEL) {
+			installEnv := &adiiov1alpha1.InstallEnv{}
+			if err := r.Get(ctx, types.NamespacedName{Name: value, Namespace: bmh.Namespace}, installEnv); err != nil {
+				return reconcileError{client.IgnoreNotFound(err)}
+			}
+
+			if installEnv.Status.ISODownloadURL == "" {
+				// the image has not been created yet, try later.
+				r.Log.Infof("Image URL for InstallEnv (%s/%s) not available yet. Retrying reconcile for BareMetalHost  %s/%s",
+					installEnv.Namespace, installEnv.Name, bmh.Namespace, bmh.Name)
+				return reconcileRequeue{time.Minute}
```

Comment:
It would be better to watch the InstallEnv resource and queue this resource when it has an event. You can use a predicate or other logic to filter out events when the URL is not set.

Reply:
> It would be better to watch the InstallEnv resource and queue this resource when it has an event. You can use a predicate or other logic to filter out events when the URL is not set.

I agree! I will leave this for a follow-up PR, tho.